Zend/zend_vm_execute.h

```diff
@@ -1779,6 +1772,27 @@ static ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL ZEND_HANDLE_EXCEPTION_SPEC_HANDLER(
 
 	cleanup_unfinished_calls(execute_data, throw_op_num);
 
+	if (throw_op->result_type & (IS_VAR | IS_TMP_VAR)) {
+		switch (throw_op->opcode) {
+			case ZEND_ADD_ARRAY_ELEMENT:
+			case ZEND_ROPE_ADD:
+				break; /* exception while building structures, live range handling will free those */
+
+			case ZEND_FETCH_CLASS:
+			case ZEND_DECLARE_CLASS:
+			case ZEND_DECLARE_INHERITED_CLASS:
+			case ZEND_DECLARE_ANON_CLASS:
+			case ZEND_DECLARE_ANON_INHERITED_CLASS:
+				break; /* return value is zend_class_entry pointer */
+
+			case ZEND_JMP_SET:
+				break; /* takes care of handling itself */
```

Comment:
> Yep, but that exception check contains HANDLE_EXCEPTION(); which will override the value with IS_UNDEF, while we need to destroy the value we have in it currently.

Ahhh, I get it now, thanks. This problem also exists in other places. For example this leaks:

```php
<?php

class Foo implements IteratorAggregate {
    public function getIterator() {
        return new ArrayIterator([]);
    }
    public function __destruct() {
        throw new Error;
    }
}

try {
    foreach (new Foo as $x);
} catch (Error $e) {
    echo $e, "\n";
}
```